### PR TITLE
Refactor/Scheduler Framework

### LINKED ISSUE #56: [TODO] switch of enable/disable short-term schedule
The implementation of ts-fsrs:
- https://github.com/open-spaced-repetition/ts-fsrs/pull/100

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
close #56 

Major change.

### CODE REVIEW

File: src/fsrs/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    @abstractmethod
+    def review_state(self, rating: Rating) -> SchedulingInfo:
+        pass
+
+    def _init_seed(self):
+        time = int(self.now.timestamp() * 1000)
+        reps = self.current.reps
+        mul = self.current.difficulty * self.current.stability
+        self.parameters.seed = f"{time}_{reps}_{mul}"
```
Comment (ishiko732) [inline src/fsrs/scheduler.py:79]: I think this section of code should be submitted in the next PR.

**Comment by joshdavham** [comment]
@L-M-Sherlock @ishiko732 

Before I start reviewing this PR, I have an important question: Are we okay with introducing major breaking changes this often?

V2 was released in April and V3 was released three weeks ago and now we're already discussing a V4. On top of that, the last couple of weeks, I've been working on a list of changes/improvements I'd like to potentially implement into this library which could result in more breaking changes in the future (V5, V6, V7, etc).

Granted, we likely don't have many users who are currently relying on this library and this is evidenced in part by how infrequently we see new issues being created by other people. So introducing frequent breaking changes won't negatively impact that many people.

I see two possibilities going forward:
1. We continue to regularly introduce major changes.
2. We spend some time now planning the future development of this project so that we won't have to introduce as many breaking major changes.

Pros for 1:
- We iterate faster
- More flexibility
- Learn faster

Cons for 1:
- Library is significanly harder for our users
- Project is harder to maintain

--

Pros for 2:
- Library is easier for our users
- Project is easier to maintain

Cons for 2:
- We iterate slower
- This ultimately won't immunize us from more breaking changes in the future, but it would hopefully reduce their number

If we decide on 1, I can start reviewing the PR this weekend.
And if we decide on 2, I can start working on a draft detailing what we might want our API to look like in the future and some of the other changes we'd like to see for this project going forward (including the changes here).

As for my opinion on what we should do... normally, if we were building an application (rather than I library), I'd recommend we do 1, but because we're building a library that other developers build on, I think I'm gonna vote for 2.

Let me know what y'al think

**Comment by L-M-Sherlock** [comment]
Major breaking changes are common when the lib in its early stage. So it's OK, especially the breaking changes have been implemented in ts-fsrs and go-fsrs.

Planning is great. I support to plan the development and create something like roadmap. I have added TODO in the issue before I work on this PR.

**Review by ishiko732 — APPROVED** [review]
LGTM

**Comment by joshdavham** [comment]
@L-M-Sherlock 

Alrighty, I'll start taking a look at this PR then. I'll probably also send you a rough outline in the next couple of weeks of some of the changes I'd like to make to this library in the future.

**Comment by joshdavham** [comment]
@L-M-Sherlock 

Alright I went and spent the last couple hours reviewing this PR and, unfortunately, I don't agree with this refactoring.

If I had to summarize, I'd say: While I support bringing py-fsrs to feature parity with ts-fsrs, **I don't support turning py-fsrs into ts-fsrs**. I'm sure ts-fsrs is just fine, but I don't think making py-fsrs equal to ts-fsrs in its implementation should be a goal.

I'll illustrate just a couple of my concerns below:

1. Parameters
Previously, you could initialize the FSRS scheduler by providing the weights, request retention and maximum interval to it directly. Now if you'd like to initialize a custom scheduler, you must first initialize a Parameters object. I can't see any good reason to implement it this way other than that's how ts-fsrs does it. I believe this adds unecessary encapsulation and complexity to something that was otherwise straightforward. If we look at other popular python libraries, e.g., [sklearn's RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html), we find that, regardless of there being many possible parameters to intialize an object with, there's no need to encapsulate them e.g. 'class RandomForestParameters'.
2. review_card -> next
Similar to the above, I don't understand what the motivation was to make a change here other than because that's what ts-fsrs does. 'review_card' as a name is much more semantic and clear than 'next'.
Furthermore, I don't see what was wrong wrong with review_card returning a tuple rather than a SchedulingInfo object other than that's because how ts-fsrs does it. If we look again at something like sklearn's RandomForestClassifier [predict_proba method](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.predict_proba), we see that it just returns a numpy array or a list of numpy arrays. There's no need to encapsulate the output with something like 'class Probabilities'.
3. preview
I spent some time trying to figure out what the preview method in the AbstractSheduler was meant for, other than replicating ts-fsrs, but it looks like it's trying to replace the SchedulingCard's schedule method. Despite this, SchedulingCards has not yet been removed in this refactoring. Additionally, while I'm sure having something like repeat/preview makes sense for code that is directly integrated into anki like fsrs-rs, I don't think it's needed in this implementation as, if a developer wants to provide preview cards, they can simply call review/next on a card with various ratings themselves.
4. Overuse of encapsulation
In case this point was lost above, I'll reiterate: I believe there is too much unecessary encapsulation in this refactoring. I get that that's how ts-fsrs does it, but I don't believe it's necessary the way it is. I worry that such encapsulation will make this repo both harder to maintain and to develop on top of.

I look forward to your response, but as it currently stands, I don't agree with this refactoring.

**Comment by ishiko732** [comment]
Sorry about the confusion. I agree that py-fsrs can have its own unique design while achieving feature parity. However, I want to clarify a few things:

1. In ts-fsrs, it’s not mandatory to initialize an FSRS instance with a Parameter object. The `generatorParameters` method is provided to allow developers to have a complete set of FSRS parameters when initializing the `User/Deck` entities, but in reality, you can create an FSRS instance by passing only some of the parameters.

2. The `next` method in ts-fsrs allows for passing the `(Card, Date, Grade)` parameters. The idea behind it is to express the next state of the card. Prior to version `4.0.0`, ts-fsrs did not have a `review_card` method, while py-fsrs had it in earlier versions. Therefore, I believe this refactoring can keep `review_card` in place. @L-M-Sherlock @joshdavham 

3. `AbstractScheduler` is meant to replace `SchedulingCard`. In ts-fsrs, SchedulingCard has already been deprecated and removed, so this PR should remove SchedulingCard as well. @L-M-Sherlock 

4. The encapsulation in ts-fsrs is aimed at decoupling short-term and long-term scheduling and decoupling the Scheduler itself to avoid conditionally executing actions based on parameters. ts-fsrs will soon export AbstractScheduler, BasicScheduler, and LongTermScheduler to allow developers to customize scheduling. 
> For example, for `new->learning` transitions, one might use BasicScheduler, and when a card in the `State.review` stage receives an `Again` rating, LongTermScheduler would be called, etc. I don’t want to rely on the `enable_short_term` parameter to determine whether a specific method should be executed or to modify the card’s state accordingly.

**Review by ishiko732 — CHANGES_REQUESTED** [review]
remove the `SchedulingCard` class

**Comment by L-M-Sherlock** [comment]
OK. I agree it's unnecessary to keep py-fsrs consistent with ts-fsrs completely. But I hope py-fsrs could support #56 and #60, whatever the way it implements. And as @ishiko732 mentioned, ts-fsrs's design has some advantages, but my PR drops some of them because I don't know how to do the same thing in Python.

**Comment by asukaminato0721** [comment]
> I don't know how to do the same thing in Python.

maybe try <https://typing.readthedocs.io/en/latest/spec/protocol.html#protocol-members>

image

notice the warning :)

```py
from typing import Protocol


class Methods(Protocol):
    def A(self): ...
    def B(self): ...
    def C(self): ...


def doit(d: Methods):
    d.A()
    d.B()
    d.C()


class Runner(Methods): ...


r = Runner()
doit(r)
```

**Comment by joshdavham** [comment]
@L-M-Sherlock @ishiko732 

Thanks you guys for your input

As for next steps, I suggest two alternatives:
1. We try to salvage this PR by refactoring this refactoring
2. We plan a rough roadmap now and get back to #56 and #60 later

Both 1. and 2. are doable, but I worry that if we do 1., we'll hit another stalemate. If we decide on 2., I can put something simple together and get your guys's feedback on it. I'm thinking we list out some of the changes we'd like to see going forward, agree on an API for this library and then decide roughly in which order we'd like to implement these changes.

Let me know what y'all think

**Comment by L-M-Sherlock** [comment]
OK. Let's do 2.
